dotCMS/src/main/java/com/dotcms/api/content/VanityUrlFactoryImpl.java

```diff
@@ -0,0 +1,226 @@
+				throw new DotStateException("Contentlet : " + con.getInode() + " is not a Vanity Url",e);
+			}
+		}else{
+			throw new DotStateException("Contentlet is null");
+		}
+
+		VanityUrl vanityUrl;
+		try {
+			vanityUrl = (VanityUrl) CacheLocator.getVanityURLCache().get(VanityUrlUtil.sanitizeKey(con));
+		} catch (DotDataException | DotRuntimeException | DotSecurityException e1) {
+			throw new DotStateException(e1);
+		}
+		if(vanityUrl!=null){
+			return vanityUrl;
+		}
+		vanityUrl = new VanityUrl();
+		vanityUrl.setStructureInode(con.getContentTypeId());
+		try {
+			contentletAPI.copyProperties((Contentlet) vanityUrl, con.getMap());
+		} catch (Exception e) {
+			throw new DotStateException("Page Copy Failed", e);
+		}
+		vanityUrl.setHost(con.getHost());
+		if(UtilMethods.isSet(con.getFolder())){
+			try{
+				Identifier ident = APILocator.getIdentifierAPI().find(con);
+				User systemUser = APILocator.systemUser();
+				Host host = APILocator.getHostAPI().find(con.getHost(), systemUser , false);
+				Folder folder = APILocator.getFolderAPI().findFolderByPath(ident.getParentPath(), host, systemUser, false);
+				vanityUrl.setFolder(folder.getInode());
+			}catch(Exception e){
+				vanityUrl=new VanityUrl();
+				Logger.warn(this, "Unable to convert contentlet to Vanity URL " + con, e);
+			}
+		}
+
+		try {
+			CacheLocator.getVanityURLCache().add(VanityUrlUtil.sanitizeKey(vanityUrl),vanityUrl);
+		} catch (Exception e) {
+
+		}
+
+		return vanityUrl;
+	}
+
+	@Override
+	public void publish(VanityUrl vanityUrl, User user, boolean respectFrontendRoles)
+			throws DotContentletStateException, DotDataException, DotSecurityException {
+
+		contentletAPI.publish(vanityUrl, user, respectFrontendRoles);
+	}
+
+	@Override
+	public void unpublish(VanityUrl vanityUrl, User user, boolean respectFrontendRoles)
+			throws DotContentletStateException, DotDataException, DotSecurityException {
+		contentletAPI.unpublish(vanityUrl, user, respectFrontendRoles);
+	}
+
+	@Override
+	public void archive(VanityUrl vanityUrl, User user, boolean respectFrontendRoles)
+			throws DotContentletStateException, DotDataException, DotSecurityException {
+		contentletAPI.archive(vanityUrl, user, respectFrontendRoles);
+	}
+
+	@Override
+	public void unarchive(VanityUrl vanityUrl, User user, boolean respectFrontendRoles)
+			throws DotContentletStateException, DotDataException, DotSecurityException {
+		contentletAPI.unarchive(vanityUrl, user, respectFrontendRoles);
+	}
+
+	@Override
+	public void delete(VanityUrl vanityUrl, User user, boolean respectFrontendRoles)
+			throws DotContentletStateException, DotDataException, DotSecurityException {
+		contentletAPI.delete(vanityUrl, user, respectFrontendRoles);
+	}
+
+	@Override
+	public VanityUrl save(VanityUrl vanityUrl, User user, boolean respectFrontendRoles) throws DotDataException, DotSecurityException{
+		if(vanityUrl != null){
+			removeFromVanityURLCache(vanityUrl);
```

Comment:
The remove from cache should be the last thing you do b/c it can get reloaded between this remove and the save.